ddl/ddl_api.go

```diff
@@ -2305,6 +2338,79 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 	return errors.Trace(err)
 }
 
+// AddColumns will add multi new columns to the table.
+func (d *ddl) AddColumns(ctx sessionctx.Context, ti ast.Ident, specs []*ast.AlterTableSpec) error {
+	schema, t, err := d.getSchemaAndTableByIdent(ctx, ti)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check all the columns at once.
+	addingColumnNames := make(map[string]bool)
+	dupColumnNames := make(map[string]bool)
+	for _, spec := range specs {
+		for _, specNewColumn := range spec.NewColumns {
+			if !addingColumnNames[specNewColumn.Name.Name.L] {
+				addingColumnNames[specNewColumn.Name.Name.L] = true
+			} else {
+				if spec.IfNotExists {
+					dupColumnNames[specNewColumn.Name.Name.L] = true
+					continue
+				}
+				return errors.Trace(infoschema.ErrColumnExists.GenWithStackByArgs(specNewColumn.Name.Name.O))
+			}
+		}
+	}
+	columns := make([]*table.Column, 0)
```

Comment:
Hi, I changed this, and I add 3 lines code after loop:
```
	// Since the newColumnsCount can be less than len(addingColumnNames)
	columns = columns[:newColumnsCount]
	positions = positions[:newColumnsCount]
	offsets = offsets[:newColumnsCount]
```